src/program.rs

```diff
@@ -104,9 +102,87 @@ impl<App: Application> Program<App> {
     ) -> Result<Program<App>, Error> {
         let typed_program = untyped_program.type_check()?;
         let witness_program = typed_program.decode_witness(iter)?;
-        let finalized_program = compress_and_finalize(witness_program);
+        let finalized_program = witness_program.finalize();
         Ok(finalized_program)
     }
+
+    /// Check if the program has maximal sharing.
+    ///
+    /// This imposes the following conditions:
+    /// 1. For hidden nodes, their hash must be unique in the program.
+    /// 2. For non-hidden nodes, the triple of their IMR, source type and target type
+    ///    must be unique in the program.
+    pub fn has_maximal_sharing(&self) -> bool {
+        let mut seen_hashes = HashSet::new();
+        let mut seen_keys = HashSet::new();
+
+        for node in &self.nodes {
+            if let Term::Hidden(h) = node.term() {
+                if seen_hashes.contains(h) {
+                    return false;
+                } else {
+                    seen_hashes.insert(h);
+                }
+            } else {
+                let primary_key = (node.imr, node.source_ty().clone(), node.target_ty().clone());
```

Comment:
In c010975498dee1db8d1654ea7b2fd47bf78a653b:

I'd prefer you use `Arc::clone` rather than `.clone` to make it explicit that you're just cloning a rc rather than a full structure (and this will also prevent future changes from making these `clone`s become expensive).

The codebase is not very consistent about doing this, but that's because I didn't know that you could call `Arc::clone` back when I first wrote it :)